### ISSUE TITLE
Strings and Symbols with the same characters should compare equal

### ISSUE THREAD

**Issue by blairmcg** [dolphinsmalltalk/Dolphin#518]
Historically Strings and Symbols with the same character sequence have not compared equal in Smalltalk, nor had the same hash. It may have been the need to provide a quick hash for Symbols (since these are often used to key into Dictionaries) that motivated this choice in Smalltalk-80, but it doesn't really make sense today and causes some awkwardness elsewhere in the class library. It also fails against the "least surprise" test.
There are some subtle compatibility issues that might arise (e.g. cases where strings and symbols are stored together in the same set, or as keys in the same dictionary), but these don't seem worth worrying about.

**Comment by jgfoster** [comment]
According to the ANSI standard, "two objects are considered equivalent if they can be used interchangeably." Also, they need to be equivalent in the other direction and there may be code that assumes that Symbol implements #'=' (equivalent) as #'==' (identical). I know that I was initially confused by the inequality but I was taught that this was "right" and I'd be hesitant to abandon the Standard (and compatibility with other dialects) without good reasons.

**Comment by blairmcg** [comment]
Fair comment, but actually there is no dialect compatibility on this any more. It is the case that #abc = 'abc' (and vice versa) in Pharo, and although I was surprised by it initially, having also come to justify to myself that #abc != 'abc' was right, I came around to thinking that actually it really isn't right, it's an optimisation to define equality as identity when there are other value equivalent objects that are not identical. That's hijacking #= to mean something else, and not something that correct code should be relying on. As I dimly recall (or possibly misremember), I don't think the standard was intended to preclude either, and the wording doesn't; a Symbol is just an interned String, and is a subclass of String, so they are interchangeable within reason.

**Comment by rko281** [comment]
Coincidentally I was musing on this issue this week as it bit me whilst porting some Pharo code (Zinc) which makes the #abc = 'abc' assumption in multiple places. 

My gut instinct is Symbols should only compare equal to themselves, and if one regards Symbols as abstract objects rather than a specialisation of String I think this is correct. However Symbol is usually (always?) implemented as a subclass of String, the Dolphin class comment states "Symbol is the class of <String>s which are guaranteed to be unique throughout the system" and the ANSI standard defines Symbols as "strings that are identity objects". 

Taking this together with the added convenience when porting frameworks I'm persuaded to accept Pharo's more relaxed semantics.

**Comment by blairmcg** [comment]
Just to add a bit of further weight from the ANSI standard; the <symbol> protocol is a sub-protocol of <readableString>, which is defined as follows:

> Provides protocol for string operations such as copying, comparing, replacing, converting, indexing, and matching. All objects that conform to the protocol <readableString> are _comparable_.

As a subprotocol, objects that conform to <symbol> must implicitly conform with <readableString> (unless refined, which is not the case), and therefore be _comparable_ with all other <readableStrings>.

_Comparable_ is defined as:

> Two objects are comparable if there is an ordering defined between them.

Although this isn't entirely clear, the example given is of all types of Number being comparable by value. If there is an ordering relationship, then there will be some cases where one object is less, some greater, and some equal.

It probably wasn't intentional, but taking all that together the ANSI standard does state that strings and symbols should be comparable, and so implicitly they should also compare as equal when they represent the same sequence of characters :-)

**Comment by jgfoster** [comment]
ANSI 5.7.8.2 specifies that for #= comparisons of a <sequenceReadableCollection>, "The receiver and operand are instances of the same class." If Dolphin doesn't enforce that, it should be noted in a comment with a justification (perhaps a reference to this discussion).

**Comment by blairmcg** [comment]
ANSI 5.7.8.2 cannot (sensibly) be enforced by any Smalltalk that has more than one String class, e.g. in VW (typing this off the top of my head):
```
'abc' asFourByteString = 'abc' asByteString. "true"
'abc' asFourByteString class == 'abc' asByteString class "false"
```
That section of the standard is either in error, or there should be a refinement of `#=` in `<readableString>`. Given that the standard is generally based on the idea of protocols, and not classes, it also goes against the general philosophy to define an equality method so tightly as to require that two objects conforming to the _same protocol_ be of the _same class_.

If Smalltalk had continued to receive mainstream interest after the standard was published, then this is one of the errors that might have been revised in a subsequent publication of the standard, but as it is we are stuck with a first cut that has never been updated.

**Comment by blairmcg** [comment]
I checked an actual copy of the standard rather than a draft, just in case it was edited for publication, but it is the same. I did notice that in the wording for `sameAs:` in <readableString> it says "The message differs from the `#=` message because two strings which are not equal can collate the same, and because the receiver and the operand _do not need to conform to the same protocols_". That suggests to me that the wording of 5.7.8.2 was supposed to refer to protocols rather than classes, but that may have caused conflict with existing implementations at the time.
